nomad/state/state_store.go

```diff
@@ -896,15 +905,17 @@ func (s *StateStore) ScalingEventsByJob(ws memdb.WatchSet, namespace, jobID stri
 // UpsertNode is used to register a node or update a node definition
 // This is assumed to be triggered by the client, so we retain the value
 // of drain/eligibility which is set by the scheduler.
-func (s *StateStore) UpsertNode(msgType structs.MessageType, index uint64, node *structs.Node) error {
+func (s *StateStore) UpsertNode(msgType structs.MessageType, index uint64, node *structs.Node, opts ...NodeUpsertOption) error {
```

Comment:
I don't know if a variadic here is the best option. I picked this for a few reasons:
1. `StateStore.UpsertNode()` is called in several places (mostly tests, but still), so we will need to update it everywhere (kind of a lazy argument 😅)
2. Prevent random boolean arguments that may be hard to know what they represent from callers side.

Reasons I don't like it:
1. It's kind of weird since no other function here uses this pattern.
2. It only represents a boolean, so it may be too restrictive to expand this to other use cases.

One alternative implementation I had at some point was to have two methods: `UpserNode()`, which doesn't create a node pool, and `UpsertNodeWithNodePool()` which always creates a node pool if necessary. Then it's up to the caller (the FSM in this case) to call the appropriate function for their use case.

Reply:
Good points, thanks!